software/webapp/src/main/java/org/apache/brooklyn/entity/webapp/JavaWebAppMigrateEffector.java

```diff
@@ -0,0 +1,115 @@
+                return null;
+            }
+        });
+    }
+
+    private static void migrate(EntityInternal entity, String locationSpec) {
+        if (ServiceStateLogic.getExpectedState(entity) != Lifecycle.RUNNING) {
+            // Is it necessary to check if the whole application is healthy?
+            throw new RuntimeException("The entity needs to be healthy before the migration starts");
+        }
+
+        if (entity.getParent() != null && !entity.getParent().equals(entity.getApplication())) {
+            /*
+             * TODO: Allow nested entites to be migrated
+             * If the entity has a parent different to the application root the migration cannot be done right now,
+             * as it could lead into problems to deal with hierarchies like SameServerEntity -> Entity
+             */
+
+            throw new RuntimeException("Nested entities cannot be migrated right now");
+        }
+
+        // Retrieving the location from the catalog.
+        Location newLocation = Entities.getManagementContext(entity).getLocationRegistry().resolve(locationSpec);
+
+        // TODO: Find a better way to check if you're migrating an entity to the exactly same VM. This not always works.
+        for (Location oldLocation : entity.getLocations()) {
+            if (oldLocation.containsLocation(newLocation)) {
+                LOG.warn("You cannot migrate an entity to the same location, the migration process will stop right now");
+                return;
+            }
+        }
+
+        LOG.info("Migration process of " + entity.getId() + " started.");
+
+        // When we have the new location, we free the resources of the current instance
+        entity.invoke(Startable.STOP, MutableMap.<String, Object>of()).blockUntilEnded();
+
+        // Clearing old locations to remove the relationship with the previous instance
+        entity.clearLocations();
+        entity.addLocations(Lists.newArrayList(newLocation));
+
+        // Starting the new instance
+        entity.invoke(Startable.START, MutableMap.<String, Object>of()).blockUntilEnded();
+
+        // Refresh all the dependent entities
+
+        for (Entity applicationChild : entity.getApplication().getChildren()) {
+            // TODO: Find a better way to refresh the application configuration.
+            // TODO: Refresh nested entities or find a way to propagate the restart properly.
+
+            // Restart any entity but the migrated one.
+            if (applicationChild instanceof Startable) {
+                if (entity.equals(applicationChild)) {
+                    // The entity is sensors should rewired automatically on stop() + restart()
+                } else {
+                    // Restart the entity to fetch again all the dependencies (ie. attributeWhenReady ones)
+                    ((Startable) applicationChild).restart();
```

Comment:
Here's a really simple initialiser that does the basic job. It should do more error checking around the state of the load balancer and the entities whose hostnames have changed. YAML blueprints would include it on whatever needed to respond to a migrate effector:

```
import org.apache.brooklyn.api.entity.EntityInitializer;
import org.apache.brooklyn.api.entity.EntityLocal;
import org.apache.brooklyn.api.sensor.SensorEvent;
import org.apache.brooklyn.api.sensor.SensorEventListener;
import org.apache.brooklyn.core.entity.Attributes;
import org.apache.brooklyn.core.entity.lifecycle.Lifecycle;
import org.apache.brooklyn.entity.proxy.LoadBalancer;

public class LoadBalancerReconfigurer implements EntityInitializer {

    @Override
    public void apply(EntityLocal entity) {
        if (entity instanceof LoadBalancer) {
            LoadBalancer lb = (LoadBalancer) entity;
            entity.subscribeToMembers(
                    entity.config().get(LoadBalancer.SERVER_POOL),
                    Attributes.HOSTNAME,
                    new RefreshConfigSensorEventListener(lb));
        } else {
            // log message that this initialiser only works with load balancers.
        }
    }

    private static class RefreshConfigSensorEventListener implements SensorEventListener<String> {
        final LoadBalancer lb;

        private RefreshConfigSensorEventListener(LoadBalancer lb) {
            this.lb = lb;
        }

        @Override
        public void onEvent(SensorEvent<String> event) {
            Lifecycle lbState = lb.sensors().get(Attributes.SERVICE_STATE_ACTUAL);
            if (Lifecycle.RUNNING.equals(lbState)) {
                lb.update();
            } else {
                // log not updating when in state
            }
        }
    }
}
```